hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/AltoChannelConnector.java

```diff
@@ -0,0 +1,215 @@
+                + connection.toLessDescriptiveString());
+
+        Channel channel = null;
+        try {
+            Address address = new Address(host, port);
+            channel = channelCreator.create(address, connection, channelInitializer);
+            writeAuthenticationBytes(channel);
+            onSuccessfulChannelConnection(channel, index);
+        } catch (Exception e) {
+            logger.warning("Exception during the connection to attempt to Alto channel on port "
+                    + port + " for " + connection.toLessDescriptiveString() + ": " + e, e);
+            onFailure(channel);
+        }
+    }
+
+    private void writeAuthenticationBytes(Channel channel) {
+        // first thing we need to send is the clientUuid so this new socket can be connected
+        // to the connection on the member.
+        ClientMessage clientUuidMessage = ClientMessage.createForEncode();
+        ClientMessage.Frame initialFrame = new ClientMessage.Frame(
+                new byte[UUID_SIZE_IN_BYTES], UNFRAGMENTED_MESSAGE);
+        encodeUUID(initialFrame.content, 0, clientUuid);
+        clientUuidMessage.add(initialFrame);
+        if (!channel.write(clientUuidMessage)) {
+            throw new HazelcastException("Cannot write authentication bytes to the channel");
+        }
+    }
+
+    private void onSuccessfulChannelConnection(Channel channel, int index) {
+        synchronized (altoChannels) {
+            if (connectionFailed()) {
+                // It might be the case that the connection or any
+                // of the channels are failed after this channel
+                // is established. We need to close this one as well
+                // to not leak any channels.
+                logger.warning("Closing the Alto channel " + channel + " for "
+                        + connection.toLessDescriptiveString()
+                        + " as one of the connections are failed.");
+                onFailure(channel);
+                return;
+            }
+
+            altoChannels[index] = channel;
+        }
+
+        logger.info("Successfully connected to Alto channel " + channel + " for "
+                + connection.toLessDescriptiveString());
+
+        if (remaining.decrementAndGet() == 0) {
+            connection.setAltoChannels(altoChannels);
+
+            // If the connection is alive at this point, but
+            // closes afterward, the channels will be cleaned up
+            // properly in the connection's close method, because
+            // we have already written the channels.
+
+            // If the connection is not alive at this point, the channels
+            // might or might not be closed, depending on the order of the
+            // close and setAltoChannels calls. We will close channels
+            // if the connection is not alive here, just in case, as it is
+            // OK to call close on already closed channels.
+            if (!connection.isAlive()) {
```

Comment:
If you synchronized on `TcpClientConnection.altoChannels` on method `TcpClientConnection.innerClose` and if you also use the same synchronization when you do the `connection.setAltoChannels(altoChannels);` then you would not need this second path for the `closeAllChannels ` method, right?

Reply:
Well, I still would need to check the state of the connection, unless I put all of this method's content into the synchronized block. I think this approach is simple and working, and I don't want to change it